### ISSUE TITLE
Upgrade `AbstractBackend`

### ISSUE THREAD

**Issue by alecandido** [qiboteam/qibo#1222]
I would make it a little less abstract, implementing most of the operations that are almost never redefined.

This would avoid the common pattern of most of the backends subclassing the `NumpyBackend` (that always acts as a proxy for the abstract one).

In a sense, we could merge the two, since we only need one. But for the time being I'd rather keep the `NumpyBackend`, implementing just `execute_circuit` and things that could be specific to it, as an example of how to implement a very simple backend.

**Comment by alecandido** [comment]
> > I am not that familiar with all the listed backends, so not much to comment, however I fully support the idea of avoiding a design where everybody inherits from `NumpyBackend`. I have the impression that this common inheritance pattern exists because `AbstractBackend` defines too much functionality, some of which can be implemented elsewhere and not as part of backends. So I would suggest to look into this from this angle as well.
> 
> I partially agree with this, but on the other side, I'm not too much against the broad scope of backend.
> 
> What I have in mind is that, if we trimmed down the functionalities of the backend, but part of those functions could have a different implementation at least for one backend, then we would end up implementing a backend mechanism once more somewhere else.
> 
> So, in my mind, the backend should expose every function that could have multiple implementations (and it's used for Qibo operations). Where multiple does not mean exactly the number of backends available, but just > 1.
> 
> However, since most of these functions are frequently not performance-critical, and often good enough in their NumPy flavor, I'd just provide a default implementation of them.
> (I.e. I'm open to implement in the `AbstractBackend` everything but `execute_circuit`, in principle - but there could be something else that is also worth to keep abstract)
> 

_Originally posted by @alecandido in https://github.com/qiboteam/qibo/issues/1117#issuecomment-1947872228_

**Comment by alecandido** [comment]
> some methods are not actually methods (i.e. they do not use `self`).

Agreed, they could be at least static (though we could discuss if using classes is the best pattern at all for implementing backends)

> some methods do use `self`, but it looks artificial - e.g. the `NumpyBackend` class stores `np` as instance variable, and then methods refer to `self.np` instead of `np`, which accounts for a significant amount of usages of `self`.

This is an issue on its own: I believe that `self.np` is used inconsistently. The idea was to leverage the various NumPy like interfaces, through the module object. [Autoray](https://github.com/jcmgray/autoray/) is going very deep in this direction, and indeed it has almost no interface on its own, just a `.do()` function that dispatches whatever operation.
However, this is not what Qibo is doing, and I believe we could have an edge over Autoray, because of deeper exploitation of frameworks' peculiarities. But I'm not sure if in this scheme `self.np` fits at all (though it could be a way to lighten fully repeated implementations).

It is worth mentioning in this context that NumPy is making remarkable efforts towards API standardization:
https://numpy.org/doc/stable/reference/arrays.interface.html
https://numpy.org/doc/stable/reference/array_api.html
(plus Python buffer protocol, plus DLPack, plus ...)

> a lot of methods make sense only for simulator backends, so carrying them around for real backends makes things confusing. In some cases might even be contradicting.

Yes, we definitely need a careful review. We might end up supporting a union of methods, but they don't need to be defined all in the `AbstractBackend`, we could have specializations.

> Some of these things will perhaps be better off implemented somewhere else. We do not need to implement them in the base class to facilitate code sharing. But as I said, I am not that familiar with the delicacies in the backend implementations, so i don't have any more concrete proposal at this point. I can take a deeper look if needed.

Agreed. Let's start doing something, and we'll go into further details.

**Comment by renatomello** [comment]
@alecandido is this issue still relevant after the backend refactoring?

**Comment by alecandido** [comment]
Hi @renatomello: no, I would say that the backend refactoring exactly implemented what was described in this issue. I'm closing it.

Thanks for your massive effort!
